CONTRIBUTING.md

```diff
@@ -0,0 +1,161 @@
+# Contributing
+
+- [Contributing](#contributing)
+  - [Overview](#overview)
+    - [Pull request rules](#pull-request-rules)
+  - [Development Procedure](#development-procedure)
+    - [Testing](#testing)
+    - [Pull Requests](#pull-requests)
+    - [Requesting Reviews](#requesting-reviews)
+    - [Updating Documentation](#updating-documentation)
+  - [Dependencies](#dependencies)
+  - [Protobuf](#protobuf)
+  - [Branching Model and Release](#branching-model-and-release)
+    - [PR Targeting](#pr-targeting)
+
+## Overview
+
+This document codifies rules that must be followed when contributing to
+Babylon node repository.
+
+### Pull request rules
+
+Every pull request must be easy to review. To make it possible:
+
+ 1. **Each pull request must do _one thing_**. It must be very clear what that
+    one thing is when looking at the pull request title, description, and linked
+    issues. It must also be very clear what value it ultimately aims to deliver,
+    and for which user(s).
+
+ 2. **Each pull request must be manageable in size**.
+    Self-contained pull requests that are manageable in size may target `main` directly.
+    Larger contributions though must be structured as a series of smaller pull requests
+    each building upon the previous one, all ideally tracked in a tracking issue
+    These pull requests must target a long-lived feature branch.
+    For details, see the [development procedure guidelines](#development-procedure).
+
+    **Note**: This does not necessarily apply to documentation-related changes or
+    automatically generated code (e.g. generated from Protobuf definitions). But
+    automatically generated code changes should occur within separate commits, so
+    they are easily distinguishable from manual code changes.
+
+## Development Procedure
+
+`main` must be stable, include only completed features and never fail `make test`, `make test-e2e`, or `make build/install`.
+
+Depending on the scope of the work, we differentiate between self-contained pull requests and long-lived contributions (features).
+
+**Self-contained pull requests**:
+
+* Fork the repo (core developers must create a branch directly in the Babylon repo),
+branch from the HEAD of `main`, make some commits, and submit a PR to `main`.
+* Follow branch name conventions: `{prefix}/branch-name`, where `prefix` is one of the
+standard [types](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json)
+* See [Branching Model](#branching-model-and-release) for more details.
+
+**Large contributions**:
+
+* Make sure that a feature branch is created in the repo or create one.
+  The name convention for the feature branch must be `feat/branch-name`.
+  Note that (similar to `main`) all feature branches have branch protection rules and they run the CI.
```

Comment:
How do we enforce this? Can external dev create a feature branch and apply branch protection rules?

Reply:
we can set patterns in the workflows file to run in pattern branches `feat/**` or `*/feat/**`

Reply:
>How do we enforce this? Can external dev create a feature branch and apply branch protection rules?

Hmm general flow seems that feature branches are usually reserved for bigger features, so for external contributor to go for such bigger feature it should be first discussed with core team, and core team can create such feature branch. (that is at least what Gaia does)

In general, I kind of treated external contributors with less details for now as imo more important is to have core team to agree on the new flow.